src/domain/usecases/user-monitoring/permission-fixer/RunUserPermissionUseCase.ts

```diff
@@ -0,0 +1,454 @@
+                            } else {
+                                return [];
+                            }
+                        })
+                    );
+
+                    const allExceptionsToBeIgnoredByRole: string[] = _.compact(
+                        excludedRolesByRole.flatMap(item => {
+                            if (item.active_role.id in user.userRoles) return item.ignore_role.id;
+                        })
+                    );
+                    const allValidRolesSingleListWithExceptions = _.concat(
+                        allValidRolesSingleList,
+                        allExceptionsToBeIgnoredByRole,
+                        allExceptionsToBeIgnoredByUser,
+                        allExceptionsToBeIgnoredByGroup
+                    );
+                    //the invalid roles are the ones that are not in the valid roles
+                    const allInvalidRolesSingleListFixed = allInValidRolesSingleList?.filter(item => {
+                        return allValidRolesSingleListWithExceptions.indexOf(item) == -1;
+                    });
+                    //fill the valid roles in the user  against all the possible valid roles
+                    const userValidRoles = user.userCredentials.userRoles.filter(userRole => {
+                        return (
+                            JSON.stringify(allValidRolesSingleListWithExceptions).indexOf(userRole.id) >= 0
+                        );
+                    });
+
+                    //fill the invalid roles in the user against all the possible invalid roles
+                    const userInvalidRoles = user.userCredentials.userRoles.filter(userRole => {
+                        return (
+                            JSON.stringify(allValidRolesSingleListWithExceptions).indexOf(userRole.id) ==
+                                -1 && JSON.stringify(allInvalidRolesSingleListFixed).indexOf(userRole.id) >= 0
+                        );
+                    });
+
+                    //clone user
+                    const fixedUser = JSON.parse(JSON.stringify(user));
+                    fixedUser.userCredentials.userRoles = userValidRoles;
+                    fixedUser.userRoles = userValidRoles;
+
+                    if (AllGroupMatch.length > 1) {
+                        log.debug(`Debug: User have more than 1 group ${user.id} - ${user.name}`);
+                        const userInfoRes: UserMonitoringUserResponse = {
+                            user: user,
+                            fixedUser: fixedUser,
+                            validUserRoles: userValidRoles,
+                            actionRequired: userInvalidRoles.length > 0,
+                            invalidUserRoles: userInvalidRoles,
+                            userNameTemplate: templateGroupMatch!.template.name,
+                            templateIdTemplate: templateGroupMatch!.template.id,
+                            groupIdTemplate: templateGroupMatch!.group.id,
```

Comment:
To avoid the not undefined asserts you could add a check after creating templateGroupMatch with an exception:
````typescript
if (templateGroupMatch === undefined) {
  throw new TypeError('....');
}
````